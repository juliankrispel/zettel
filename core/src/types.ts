/**
 * Container for Entities
 */
export type EntityMap = {
  [key: string]: Entity
}

/**
 * Entity ftw!
 */
export type Entity = {
  isAtomic?: boolean,
  [key: string]: any
}

/**
 * State of a document.
 * Used to create derived states such as BlockTree,
 * RawDocument or any other
 */
export type ListState = {
  value: Value,
  entityMap: EntityMap,
}

/**
 * Array of Characters
*/
export type Value = Character[]

export type Character = (TextCharacter | BlockStart | BlockEnd)

export type CharacterData = {
  styles: string[],
  entity?: string | null,
  type?: void,
}

/*
* Represents one text character.
* Contains text symbol as well as array of
* styles and entity
*/
export type TextCharacter = CharacterData & {
  char: string,
}

/**
 * Represents the beginning of a block
 */
export type BlockStart = {
  blockKey: string,
  type: 'block-start',
  entity?: string | null
}

export type TextFragment = CharacterData & {
  text: string,
  entity?: Entity
}

/**
 * Represents the end of a block, there
 * has toe be a matching 
 */
export type BlockEnd = {
  type: 'block-end'
}

/**
 * Embodies all possible changes made to doc
 */
export type Change = {
  start: number,
  end: number,
  value: Value | CharacterData
}

/**
 * History
 */
export type Changes = Change[]

/**
 * Represents raw form of a Zettel Document,
 * serializable and easier on the human eye
 */
export type RawDocument = {
  text: string,
  ranges: RawRange[],
  entityMap: EntityMap,
}

/**
 * Ranges in RawDocument
 * Used to style text and relate to entities
 */
export type RawRange = {
  offset: number,
  length: number,
  styles: string[],
  entity?: string | null,
}

export type CharacterRange = RawRange

/**
* Tree representation of content. Used for
* rendering only
*/
export type BlockTree = {
  blocks: Block[],
  entityMap: EntityMap,
}

/**
* Represents one node in a tree
* Used for rendering
*/
export type Block = {
  value: TextCharacter[],
  blockKey: string,
  blocks: Block[],
  entity?: Entity | null
}


/**
 * 
 */
export interface EditorState {
  /**
   * @method change()
   * 
   * `EditorState.change({ start: 9, end: 18, value: [] })`
   */
  change: (changes: Change[]) => EditorState;

  /**
   * @method undo()
   * 
   * `EditorState.undo()`
   */
  undo: () => EditorState;

  /**
   * @method redo()
   * 
   * `EditorState.redo()`
   */
  redo: () => EditorState;

  readonly list: ListState;
  readonly tree: BlockTree;
}