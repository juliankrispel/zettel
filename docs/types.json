[
  {
    "file": "/Users/juliankrispel/projects/zettel-new/react/src/types.ts",
    "nodes": [
      {
        "docComment": "Render",
        "kind": "TypeAliasDeclaration",
        "name": "RenderStyle",
        "type": "React.FunctionComponent<{\n  style: string,\n  children: React.ReactElement\n}>"
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/react/src/Editor.tsx",
    "nodes": [
      {
        "name": null,
        "kind": "VariableStatement",
        "docComment": "Editor Component",
        "parameters": [
          {
            "isOptional": false,
            "name": "props",
            "type": {
              "mapBlock": "(block: Block) => Block",
              "renderBlock": "RenderBlock",
              "renderChildren": "RenderChildren",
              "renderStyle": "RenderStyle",
              "renderTextFragment": "RenderTextFragment",
              "onChange": "(editorState: EditorState) => void",
              "editorState": "EditorState",
              "htmlAttrs": "Object",
              "onKeyDown": "(event: React.KeyboardEvent) => EditorState | void",
              "readOnly": "boolean"
            }
          }
        ],
        "type": "React.ReactElement"
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/types.ts",
    "nodes": [
      {
        "docComment": "Container for Entities",
        "kind": "TypeAliasDeclaration",
        "name": "EntityMap",
        "type": {}
      },
      {
        "docComment": "Entity ftw!",
        "kind": "TypeAliasDeclaration",
        "name": "Entity",
        "type": {}
      },
      {
        "docComment": "State of a document.\nUsed to create derived states such as BlockTree,\nRawDocument or any other",
        "kind": "TypeAliasDeclaration",
        "name": "ListState",
        "type": {}
      },
      {
        "docComment": "Array of Characters",
        "kind": "TypeAliasDeclaration",
        "name": "Value",
        "type": "Character[]"
      },
      {
        "docComment": "Represents the beginning of a block",
        "kind": "TypeAliasDeclaration",
        "name": "BlockStart",
        "type": {}
      },
      {
        "docComment": "Represents the end of a block, there\nhas toe be a matching",
        "kind": "TypeAliasDeclaration",
        "name": "BlockEnd",
        "type": {}
      },
      {
        "docComment": "Embodies all possible changes made to doc",
        "kind": "TypeAliasDeclaration",
        "name": "Change",
        "type": {}
      },
      {
        "docComment": "History",
        "kind": "TypeAliasDeclaration",
        "name": "Changes",
        "type": "Change[]"
      },
      {
        "docComment": "Represents raw form of a Zettel Document,\nserializable and easier on the human eye",
        "kind": "TypeAliasDeclaration",
        "name": "RawDocument",
        "type": {}
      },
      {
        "docComment": "Ranges in RawDocument\nUsed to style text and relate to entities",
        "kind": "TypeAliasDeclaration",
        "name": "RawRange",
        "type": {}
      },
      {
        "docComment": "Tree representation of content. Used for\nrendering only",
        "kind": "TypeAliasDeclaration",
        "name": "BlockTree",
        "type": {}
      },
      {
        "docComment": "Represents one node in a tree\nUsed for rendering",
        "kind": "TypeAliasDeclaration",
        "name": "Block",
        "type": {}
      },
      {
        "kind": "InterfaceDeclaration",
        "text": "export interface EditorState {\n  /**\n   * @method change()\n   * \n   * `EditorState.change({ start: 9, end: 18, value: [] })`\n   */\n  change: (changes: Change[]) => EditorState;\n\n  /**\n   * @method undo()\n   * \n   * `EditorState.undo()`\n   */\n  undo: () => EditorState;\n\n  /**\n   * @method redo()\n   * \n   * `EditorState.redo()`\n   */\n  redo: () => EditorState;\n\n  readonly list: ListState;\n  readonly tree: BlockTree;\n}"
      },
      {
        "kind": "PropertySignature",
        "text": "change: (changes: Change[]) => EditorState;"
      },
      {
        "kind": "PropertySignature",
        "text": "undo: () => EditorState;"
      },
      {
        "kind": "PropertySignature",
        "text": "redo: () => EditorState;"
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/EditorState/change.ts",
    "nodes": [
      {
        "docComment": "Represents the input and output for updates",
        "kind": "TypeAliasDeclaration",
        "name": "Update",
        "type": {}
      },
      {
        "name": "change",
        "kind": "FunctionDeclaration",
        "docComment": "takes State and Change object, return a state\nobject that has been updated with the Change, and\na Change object to undo the change",
        "type": "Update",
        "parameters": [
          {
            "isOptional": false,
            "name": "update",
            "type": "Update"
          }
        ]
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/EditorState/EditorState.ts",
    "nodes": [
      {
        "name": "getSelectedValue",
        "kind": "MethodDeclaration",
        "docComment": "gets selected value, arguments default\nto editorState.start and editorState.end",
        "type": "Value",
        "parameters": [
          {
            "isOptional": false,
            "name": "start",
            "type": "number"
          },
          {
            "isOptional": false,
            "name": "end",
            "type": "number"
          }
        ]
      },
      {
        "name": "fromJSON",
        "kind": "MethodDeclaration",
        "docComment": "creates a new EditorState from JSON format",
        "type": "EditorState",
        "parameters": [
          {
            "isOptional": false,
            "name": "json",
            "type": "RawDocument"
          }
        ]
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/getBlocksForRange.ts",
    "nodes": [
      {
        "name": "getBlocksForRange",
        "kind": "FunctionDeclaration",
        "docComment": "Gets block indeces for range (start, end)",
        "type": "BlockAndIndex[]",
        "parameters": [
          {
            "isOptional": false,
            "name": "value",
            "type": "Value"
          },
          {
            "isOptional": false,
            "name": "start",
            "type": "number"
          },
          {
            "isOptional": false,
            "name": "end",
            "type": "number"
          }
        ]
      }
    ]
  },
  {
    "file": "/Users/juliankrispel/projects/zettel-new/core/src/textToListIndex.ts",
    "nodes": [
      {
        "name": "textToListIndex",
        "kind": "FunctionDeclaration",
        "docComment": "Maps text index to list index",
        "type": "number",
        "parameters": [
          {
            "isOptional": false,
            "name": "value",
            "type": "Value"
          },
          {
            "isOptional": false,
            "name": "textIndex",
            "type": "number"
          }
        ]
      }
    ]
  }
]