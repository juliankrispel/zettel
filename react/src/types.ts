import { TextFragment, Block } from '@zettel/core'
import { CSSProperties } from 'react';

type FragmentRenderProps = {
  fragment: TextFragment
}

export type RenderStyle = React.FunctionComponent<{
  style: string,
  children: React.ReactElement
}>

export type RenderEntity = React.FunctionComponent<{
  entity: any,
  children: React.ReactElement
}>

export type RenderBlock = React.FunctionComponent<{
  block: Block,
  style: CSSProperties,
  children: React.ReactElement
}>

export type RenderChildren = React.FunctionComponent<{
  block?: Block,
  children: React.ReactElement
}>

export type RenderProps = {
  mapBlock?: (block: Block) => Block,
  renderBlock?: RenderBlock,
  renderChildren?: RenderChildren,
  renderStyle?: RenderStyle,
  renderEntity?: RenderEntity,
}  