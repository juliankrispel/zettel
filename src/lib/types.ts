/**
 * Container for Entities
 */
export type EntityMap = {
  [key: string]: any
}

/**
 * State of a document.
 * Used to create derived states such as BlockTree,
 * RawDocument or any other
 */
export type ListState = {
  value: Value,
  entityMap: EntityMap,
}

/**
* Value object. Think of this as your `value` in `<input value={value} />`,
* except instead of a string you get an array of objects.
*/
export type Value = (Character | BlockStart | BlockEnd)[]

/*
* Represents one text character.
* Contains text symbol as well as array of
* styles and entity
*/
export type Character = {
  char: string,
  styles: string[],
  entity?: string | null,
  type?: void,
}

/**
 * Represents the beginning of a block
 */
export type BlockStart = {
  type: 'block-start',
  data?: any
}

/**
 * Represents the end of a block, there
 * has toe be a matching 
 */
export type BlockEnd = {
  type: 'block-end'
}

/**
 * Embodies all possible changes made to doc
 */
export type Change = {
  start: number,
  end: number,
  value: Value
}[]

/**
 * Represents raw form of a Zettel Document,
 * serializable and easier on the human eye
 */
export type RawDocument = {
  text: string,
  ranges: RawRange[],
  entityMap: EntityMap,
}

/**
 * Ranges in RawDocument
 * Used to style text and relate to entities
 */
export type RawRange = {
  offset: number,
  length: number,
  styles: string[],
  entity?: string | null,
}

/**
* Tree representation of content. Used for
* rendering only
*/
export type BlockTree = {
  blocks: Block[],
  entityMap: EntityMap,
}

/**
* Represents one node in a tree
* Used for rendering
*/
export type Block = {
  value: Character[],
  key: string,
  blocks: Block[],
}


/**
 * 
 */
export interface EditorState {
  /**
   * @method change()
   * 
   * `EditorState.change({ start: 9, end: 18, value: [] })`
   */
  change: (change: Change) => EditorState;

  /**
   * @method undo()
   * 
   * `EditorState.undo()`
   */
  undo: () => EditorState;

  /**
   * @method redo()
   * 
   * `EditorState.redo()`
   */
  redo: () => EditorState;

  readonly list: ListState;
  readonly tree: BlockTree;
}