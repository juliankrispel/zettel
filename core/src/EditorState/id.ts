// @ts-ignore
if (process.env.NODE_ENV === 'test') {
  // @ts-ignore
  window.__ZETTEL_ID = 0
}

function id(): string {
  // @ts-ignore
  if (process.env.NODE_ENV === 'test') {
    // @ts-ignore
    window.__ZETTEL_ID++
    // @ts-ignore
    return `AAAAAA${window.__ZETTEL_ID}`
  }

  let firstPart:any = (Math.random() * 46656) | 0;
  let secondPart: any = (Math.random() * 46656) | 0;
  firstPart = ("000" + firstPart.toString(36)).slice(-3);
  secondPart = ("000" + secondPart.toString(36)).slice(-3);
  return `${firstPart}${secondPart}`;
}


export default id