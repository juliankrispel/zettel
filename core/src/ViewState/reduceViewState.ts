import { Block, Fragment, Character, ListState, ViewState, TextCharacter, BlockStart, FragmentStart, ContainerFragment } from '../types'
import getBlockByPath from './getBlockByPath'
import getChildrenByPath from './getChildrenByPath'
import getFragmentByPath from './getFragmentByPath'
import getFragmentsByPath from './getFragmentsByPath'
import createTextFragments from './createTextFragments'

type ReducerState = {
  viewState: ViewState,
  i: number,
  currentText: TextCharacter[],
  blockPath: number[],
  fragPath: number[],
}

const reducer = (state: ReducerState, char: Character, i: number) => {
  const blocks = getChildrenByPath(state.viewState, state.blockPath)
  const block = getBlockByPath(state.viewState, state.blockPath)
  const fragment = getFragmentByPath(block, state.fragPath)
  const fragments = getFragmentsByPath(block, state.fragPath)
  const type = 'type' in char ? char.type : null

  if (type == null) {
    state.currentText.push(char as TextCharacter)
    return state
  }

  if (state.currentText.length > 0) {
    if (fragment != null && 'fragments' in fragment) {
      fragment.fragments = fragment.fragments.concat(createTextFragments(state.currentText, {}))
    } else {
      block.fragments = block.fragments.concat(createTextFragments(state.currentText, {}))
    }
      
    block.value = block.value.concat(state.currentText)
    state.currentText = []
  }

  if (type === 'fragment-start') {
    let frag = char as FragmentStart
    if ('fragments' in fragment) {
      console.log('push fragment on fragment', fragment.fragments.length)
      console.log({ fragments, block })
      state.fragPath.push(fragment.fragments.length)
      fragment.fragments.push({
        fragments: [],
        data: frag.data
      })
    } else {
      console.log('push fragment on block', block.fragments.length)
      state.fragPath.push(block.fragments.length)
      block.fragments.push({
        fragments: [],
        data: frag.data
      })
    }
  } else if (type === 'fragment-end') {
    state.fragPath.pop()
  } else if (type === 'block-start') {
    let _char = char as BlockStart
    blocks.push({
      fragments: [],
      value: [],
      blocks: [],
      blockLevel: state.blockPath.length,
      blockKey: _char.blockKey,
      styles: _char.styles != null ? _char.styles : [],
    })

    state.blockPath.push(blocks.length - 1)
  } else if (type == 'block-end') {
    state.blockPath.pop()
  }


  return state
}


export default function reduceViewState (
  flat: ListState,
): ViewState {
  /**
   * Given a document contains the following tokens:
   * `TextCharacter | BlockStart | BlockEnd | FragmentStart | FragmentEnd`
   */
  const initialState: ReducerState = {
    viewState: {
      blocks: [],
      entityMap: {}
    },
    i: 0,
    blockPath: [],
    fragPath: [],
    currentText: []
  }

  return flat.value.reduce(reducer, initialState).viewState
}