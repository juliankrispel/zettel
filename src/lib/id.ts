function id(): string {
  let firstPart:any = (Math.random() * 46656) | 0;
  let secondPart: any = (Math.random() * 46656) | 0;
  firstPart = ("000" + firstPart.toString(36)).slice(-3);
  secondPart = ("000" + secondPart.toString(36)).slice(-3);
  return `${firstPart}${secondPart}`;
}


export default id