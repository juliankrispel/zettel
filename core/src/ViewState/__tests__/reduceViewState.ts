import { Value } from '../../types'
import id from '../../EditorState/id'
import reduceViewState from '../reduceViewState'

const testState: Value = [{
  blockKey: id(),
  type: 'block-start',
  styles: []
}, {
  char: 'H'
}, {
  char: 'i'
}, {
  type: 'fragment-start'
}, {
  char: 'Y'
}, {
  char: 'o'
}, {
  type: 'block-end',
}, {
  blockKey: id(),
  type: 'block-start',
  styles: []
}, {
  char: 'W'
}, {
  char: 'h'
}, {
  type: 'fragment-end'
}, {
  char: 'o'
}, {
  type: 'block-end',
}]


describe('reduceViewState', () => {
  it('should create view state with nested fragments', () => {
    const res = reduceViewState({
      value: testState,
      entityMap: {}
    })

    const expected = {
      "blocks": [
        {
          "fragments": [{
              "entity": null,
              "offset": 0,
              "text": "Hi"
            },
            {
              "entity": null,
              "offset": 0,
              "text": "Yo"
            }
          ],
          "value": [{ "char": "W" }, { "char": "h" }],
        },
        {
          "fragments": [{
              "entity": null,
              "offset": 0,
              "text": "Wh"
            },
            {
              "entity": null,
              "offset": 0,
              "text": "o"
            }
          ],
          "value": [],
          "blocks": [],
          "blockLevel": 0,
          "blockKey": "vrcged",
          "styles": []
        }
    ]
  }
  // expect(res).toContainEqual(expected)
    console.log(JSON.stringify(res, null, 2))
  })
})