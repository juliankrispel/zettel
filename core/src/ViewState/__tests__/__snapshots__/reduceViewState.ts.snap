// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduceViewState creates view state for empty block in between 1`] = `
Object {
  "blocks": Array [
    Object {
      "blockKey": "block-1",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "One",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "O",
        },
        Object {
          "char": "n",
        },
        Object {
          "char": "e",
        },
      ],
    },
    Object {
      "blockKey": "block-2",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [],
      "styles": Array [],
      "value": Array [],
    },
    Object {
      "blockKey": "block-3",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "Two",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "T",
        },
        Object {
          "char": "w",
        },
        Object {
          "char": "o",
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;

exports[`reduceViewState creates view state for nested blocks 1`] = `
Object {
  "blocks": Array [
    Object {
      "blockKey": "block-1",
      "blockLevel": 0,
      "blocks": Array [
        Object {
          "blockKey": "block-2",
          "blockLevel": 1,
          "blocks": Array [],
          "fragments": Array [
            Object {
              "text": "Two",
            },
          ],
          "styles": Array [],
          "value": Array [
            Object {
              "char": "T",
            },
            Object {
              "char": "w",
            },
            Object {
              "char": "o",
            },
          ],
        },
        Object {
          "blockKey": "block-3",
          "blockLevel": 1,
          "blocks": Array [],
          "fragments": Array [
            Object {
              "text": "Three",
            },
          ],
          "styles": Array [],
          "value": Array [
            Object {
              "char": "T",
            },
            Object {
              "char": "h",
            },
            Object {
              "char": "r",
            },
            Object {
              "char": "e",
            },
            Object {
              "char": "e",
            },
          ],
        },
      ],
      "fragments": Array [
        Object {
          "text": "One",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "O",
        },
        Object {
          "char": "n",
        },
        Object {
          "char": "e",
        },
      ],
    },
    Object {
      "blockKey": "block-4",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "Four",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "F",
        },
        Object {
          "char": "o",
        },
        Object {
          "char": "u",
        },
        Object {
          "char": "r",
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;

exports[`reduceViewState creates view state with fragments in between block boundaries 1`] = `
Object {
  "blocks": Array [
    Object {
      "blockKey": "block-1",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "Hi",
        },
        Object {
          "data": undefined,
          "fragments": Array [
            Object {
              "text": "Yo",
            },
          ],
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "H",
        },
        Object {
          "char": "i",
        },
        Object {
          "char": "Y",
        },
        Object {
          "char": "o",
        },
      ],
    },
    Object {
      "blockKey": "block-2",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "Wh",
        },
        Object {
          "text": "o",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "W",
        },
        Object {
          "char": "h",
        },
        Object {
          "char": "o",
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;

exports[`reduceViewState creates view state with nested fragments 1`] = `
Object {
  "blocks": Array [
    Object {
      "blockKey": "block-1",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "O",
        },
        Object {
          "data": undefined,
          "fragments": Array [
            Object {
              "text": "n",
            },
            Object {
              "data": Object {
                "some": "a",
              },
              "fragments": Array [
                Object {
                  "text": "e ",
                },
              ],
            },
            Object {
              "text": "TWo",
            },
          ],
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "O",
        },
        Object {
          "char": "n",
        },
        Object {
          "char": "e",
        },
        Object {
          "char": " ",
        },
        Object {
          "char": "T",
        },
        Object {
          "char": "W",
        },
        Object {
          "char": "o",
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;

exports[`reduceViewState empty block in between 1`] = `
Object {
  "blocks": Array [
    Object {
      "blockKey": "block-1",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "One",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "O",
        },
        Object {
          "char": "n",
        },
        Object {
          "char": "e",
        },
      ],
    },
    Object {
      "blockKey": "block-2",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [],
      "styles": Array [],
      "value": Array [],
    },
    Object {
      "blockKey": "block-3",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "Two",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "T",
        },
        Object {
          "char": "w",
        },
        Object {
          "char": "o",
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;

exports[`reduceViewState nested blocks 1`] = `
Object {
  "blocks": Array [
    Object {
      "blockKey": "block-1",
      "blockLevel": 0,
      "blocks": Array [
        Object {
          "blockKey": "block-2",
          "blockLevel": 1,
          "blocks": Array [],
          "fragments": Array [
            Object {
              "text": "Two",
            },
          ],
          "styles": Array [],
          "value": Array [
            Object {
              "char": "T",
            },
            Object {
              "char": "w",
            },
            Object {
              "char": "o",
            },
          ],
        },
        Object {
          "blockKey": "block-3",
          "blockLevel": 1,
          "blocks": Array [],
          "fragments": Array [
            Object {
              "text": "Three",
            },
          ],
          "styles": Array [],
          "value": Array [
            Object {
              "char": "T",
            },
            Object {
              "char": "h",
            },
            Object {
              "char": "r",
            },
            Object {
              "char": "e",
            },
            Object {
              "char": "e",
            },
          ],
        },
      ],
      "fragments": Array [
        Object {
          "text": "One",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "O",
        },
        Object {
          "char": "n",
        },
        Object {
          "char": "e",
        },
      ],
    },
    Object {
      "blockKey": "block-4",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "Four",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "F",
        },
        Object {
          "char": "o",
        },
        Object {
          "char": "u",
        },
        Object {
          "char": "r",
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;

exports[`reduceViewState should create view state with nested fragments 1`] = `
Object {
  "blocks": Array [
    Object {
      "blockKey": "block-1",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "Hi",
        },
        Object {
          "text": "Yo",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "H",
        },
        Object {
          "char": "i",
        },
        Object {
          "char": "Y",
        },
        Object {
          "char": "o",
        },
      ],
    },
    Object {
      "blockKey": "block-2",
      "blockLevel": 0,
      "blocks": Array [],
      "fragments": Array [
        Object {
          "text": "Wh",
        },
        Object {
          "text": "o",
        },
      ],
      "styles": Array [],
      "value": Array [
        Object {
          "char": "W",
        },
        Object {
          "char": "h",
        },
        Object {
          "char": "o",
        },
      ],
    },
  ],
  "entityMap": Object {},
}
`;
